File: frontend/src/types/routing.ts
/**
 * Routing types for checkpoint-based validation pipelines.
 *
 * Routing rules determine which actions to execute based on validation
 * results. They enable complex conditional logic for notifications and
 * post-validation processing.
 *
 * Features:
 * - Jinja2-based rule expressions
 * - Built-in rules (Severity, Status, Tag, etc.)
 * - Compound rules (AllOf, AnyOf, Not)
 * - Priority-based routing
 * - Context-based routing
 */

// =============================================================================
// Route Mode
// =============================================================================

/**
 * How routes are evaluated.
 */
export type RouteMode =
  | 'first_match'    // Stop at first matching route
  | 'all_matches'    // Execute all matching routes
  | 'priority'       // Execute in priority order
  | 'priority_group' // Execute highest priority group

/**
 * Human-readable labels for route modes.
 */
export const ROUTE_MODE_LABELS: Record<RouteMode, string> = {
  first_match: 'First Match',
  all_matches: 'All Matches',
  priority: 'Priority Order',
  priority_group: 'Priority Group',
}

/**
 * Descriptions for route modes.
 */
export const ROUTE_MODE_DESCRIPTIONS: Record<RouteMode, string> = {
  first_match: 'Stop at the first matching route',
  all_matches: 'Execute all matching routes',
  priority: 'Execute all matches in priority order',
  priority_group: 'Execute all routes in the highest priority group',
}

// =============================================================================
// Route Priority
// =============================================================================

/**
 * Priority levels for routes.
 */
export type RoutePriority = 'critical' | 'high' | 'medium' | 'low' | 'default'

/**
 * Numeric values for priorities.
 */
export const ROUTE_PRIORITY_VALUES: Record<RoutePriority, number> = {
  critical: 100,
  high: 75,
  medium: 50,
  low: 25,
  default: 0,
}

/**
 * Labels for priorities.
 */
export const ROUTE_PRIORITY_LABELS: Record<RoutePriority, string> = {
  critical: 'Critical (100)',
  high: 'High (75)',
  medium: 'Medium (50)',
  low: 'Low (25)',
  default: 'Default (0)',
}

// =============================================================================
// Routing Rule Types
// =============================================================================

/**
 * Available rule types.
 */
export type RoutingRuleType =
  // Simple rules
  | 'always'
  | 'never'
  | 'status'
  | 'severity'
  | 'issue_count'
  | 'pass_rate'
  | 'tag'
  | 'data_asset'
  | 'metadata'
  | 'time_window'
  | 'error'
  // Compound rules
  | 'all_of'
  | 'any_of'
  | 'not'
  // Expression rules
  | 'jinja2'
  | 'python'

/**
 * Rule type metadata.
 */
export interface RuleTypeInfo {
  type: RoutingRuleType
  label: string
  description: string
  category: 'simple' | 'compound' | 'expression'
  supportsNegate?: boolean
}

/**
 * Metadata for all rule types.
 */
export const RULE_TYPE_INFO: Record<RoutingRuleType, RuleTypeInfo> = {
  always: {
    type: 'always',
    label: 'Always',
    description: 'Always matches (default route)',
    category: 'simple',
  },
  never: {
    type: 'never',
    label: 'Never',
    description: 'Never matches (disabled route)',
    category: 'simple',
  },
  status: {
    type: 'status',
    label: 'Status',
    description: 'Match based on checkpoint status',
    category: 'simple',
    supportsNegate: true,
  },
  severity: {
    type: 'severity',
    label: 'Severity',
    description: 'Match based on issue severity levels',
    category: 'simple',
    supportsNegate: true,
  },
  issue_count: {
    type: 'issue_count',
    label: 'Issue Count',
    description: 'Match based on number of issues',
    category: 'simple',
    supportsNegate: true,
  },
  pass_rate: {
    type: 'pass_rate',
    label: 'Pass Rate',
    description: 'Match based on validation pass rate',
    category: 'simple',
    supportsNegate: true,
  },
  tag: {
    type: 'tag',
    label: 'Tag',
    description: 'Match based on checkpoint tags',
    category: 'simple',
    supportsNegate: true,
  },
  data_asset: {
    type: 'data_asset',
    label: 'Data Asset',
    description: 'Match based on data asset name pattern',
    category: 'simple',
    supportsNegate: true,
  },
  metadata: {
    type: 'metadata',
    label: 'Metadata',
    description: 'Match based on metadata values',
    category: 'simple',
    supportsNegate: true,
  },
  time_window: {
    type: 'time_window',
    label: 'Time Window',
    description: 'Match based on time of day (business hours)',
    category: 'simple',
    supportsNegate: true,
  },
  error: {
    type: 'error',
    label: 'Error',
    description: 'Match based on error occurrence',
    category: 'simple',
    supportsNegate: true,
  },
  all_of: {
    type: 'all_of',
    label: 'All Of (AND)',
    description: 'All child rules must match',
    category: 'compound',
  },
  any_of: {
    type: 'any_of',
    label: 'Any Of (OR)',
    description: 'At least one child rule must match',
    category: 'compound',
  },
  not: {
    type: 'not',
    label: 'Not',
    description: 'Invert the child rule result',
    category: 'compound',
  },
  jinja2: {
    type: 'jinja2',
    label: 'Jinja2 Expression',
    description: 'Custom Jinja2 template expression',
    category: 'expression',
  },
  python: {
    type: 'python',
    label: 'Python Expression',
    description: 'Custom Python expression (advanced)',
    category: 'expression',
  },
}

// =============================================================================
// Routing Rule Configurations
// =============================================================================

/**
 * Base rule configuration.
 */
export interface BaseRuleConfig {
  /** Rule type */
  type: RoutingRuleType
  /** Rule name */
  name?: string
  /** Rule description */
  description?: string
}

/**
 * Always rule - always matches.
 */
export interface AlwaysRuleConfig extends BaseRuleConfig {
  type: 'always'
}

/**
 * Never rule - never matches.
 */
export interface NeverRuleConfig extends BaseRuleConfig {
  type: 'never'
}

/**
 * Status rule - match by checkpoint status.
 */
export interface StatusRuleConfig extends BaseRuleConfig {
  type: 'status'
  /** Statuses to match */
  statuses: string[]
  /** Invert the match */
  negate?: boolean
}

/**
 * Severity rule - match by issue severity.
 */
export interface SeverityRuleConfig extends BaseRuleConfig {
  type: 'severity'
  /** Minimum severity to match */
  min_severity: 'critical' | 'high' | 'medium' | 'low' | 'info'
  /** Maximum severity to match */
  max_severity?: 'critical' | 'high' | 'medium' | 'low' | 'info'
  /** Minimum count of issues at this severity */
  min_count?: number
  /** Exact count of issues at this severity */
  exact_count?: number
}

/**
 * Issue count rule - match by number of issues.
 */
export interface IssueCountRuleConfig extends BaseRuleConfig {
  type: 'issue_count'
  /** Minimum issue count */
  min_issues?: number
  /** Maximum issue count */
  max_issues?: number
  /** Count type (total or by severity) */
  count_type?: 'total' | 'critical' | 'high' | 'medium' | 'low' | 'info'
}

/**
 * Pass rate rule - match by validation pass rate.
 */
export interface PassRateRuleConfig extends BaseRuleConfig {
  type: 'pass_rate'
  /** Minimum pass rate (0-100) */
  min_rate?: number
  /** Maximum pass rate (0-100) */
  max_rate?: number
}

/**
 * Tag rule - match by checkpoint tags.
 */
export interface TagRuleConfig extends BaseRuleConfig {
  type: 'tag'
  /** Tags to match (key-value pairs) */
  tags: Record<string, string | null>
  /** Match all tags (AND) or any tag (OR) */
  match_all?: boolean
  /** Invert the match */
  negate?: boolean
}

/**
 * Data asset rule - match by data asset name pattern.
 */
export interface DataAssetRuleConfig extends BaseRuleConfig {
  type: 'data_asset'
  /** Pattern to match (glob or regex) */
  pattern: string
  /** Whether pattern is regex */
  is_regex?: boolean
  /** Case sensitive matching */
  case_sensitive?: boolean
}

/**
 * Metadata rule - match by metadata values.
 */
export interface MetadataRuleConfig extends BaseRuleConfig {
  type: 'metadata'
  /** Dot-separated key path */
  key_path: string
  /** Expected value */
  expected_value?: unknown
  /** Comparison operator */
  comparator?: 'eq' | 'ne' | 'gt' | 'gte' | 'lt' | 'lte' | 'contains' | 'regex' | 'exists'
}

/**
 * Time window rule - match by time of day.
 */
export interface TimeWindowRuleConfig extends BaseRuleConfig {
  type: 'time_window'
  /** Start time (HH:MM format) */
  start_time?: string
  /** End time (HH:MM format) */
  end_time?: string
  /** Days of week (0=Monday, 6=Sunday) */
  days_of_week?: number[]
  /** Timezone */
  timezone?: string
}

/**
 * Error rule - match on error occurrence.
 */
export interface ErrorRuleConfig extends BaseRuleConfig {
  type: 'error'
  /** Error pattern to match (regex) */
  pattern?: string
  /** Invert the match (match if no error) */
  negate?: boolean
}

/**
 * AllOf rule - all child rules must match.
 */
export interface AllOfRuleConfig extends BaseRuleConfig {
  type: 'all_of'
  /** Child rules */
  rules: RoutingRule[]
}

/**
 * AnyOf rule - any child rule must match.
 */
export interface AnyOfRuleConfig extends BaseRuleConfig {
  type: 'any_of'
  /** Child rules */
  rules: RoutingRule[]
}

/**
 * Not rule - invert child rule.
 */
export interface NotRuleConfig extends BaseRuleConfig {
  type: 'not'
  /** Child rule */
  rule: RoutingRule
}

/**
 * Jinja2 rule - custom expression.
 */
export interface Jinja2RuleConfig extends BaseRuleConfig {
  type: 'jinja2'
  /** Jinja2 expression */
  expression: string
}

/**
 * Python rule - custom Python expression.
 */
export interface PythonRuleConfig extends BaseRuleConfig {
  type: 'python'
  /** Python expression */
  expression: string
}

/**
 * Union type for all routing rules.
 */
export type RoutingRule =
  | AlwaysRuleConfig
  | NeverRuleConfig
  | StatusRuleConfig
  | SeverityRuleConfig
  | IssueCountRuleConfig
  | PassRateRuleConfig
  | TagRuleConfig
  | DataAssetRuleConfig
  | MetadataRuleConfig
  | TimeWindowRuleConfig
  | ErrorRuleConfig
  | AllOfRuleConfig
  | AnyOfRuleConfig
  | NotRuleConfig
  | Jinja2RuleConfig
  | PythonRuleConfig

// =============================================================================
// Route Definition
// =============================================================================

/**
 * A route defines a rule and its associated actions.
 */
export interface Route {
  /** Route name for identification */
  name: string
  /** Routing rule to evaluate */
  rule: RoutingRule
  /** Action names to execute when rule matches */
  actions: string[]
  /** Priority for route ordering */
  priority?: RoutePriority | number
  /** Whether this route is enabled */
  enabled?: boolean
  /** Additional metadata */
  metadata?: Record<string, unknown>
  /** Stop evaluating other routes after this one matches */
  stop_on_match?: boolean
  /** Description */
  description?: string
}

/**
 * Router configuration.
 */
export interface Router {
  /** Routing mode */
  mode: RouteMode
  /** Routes */
  routes: Route[]
}

// =============================================================================
// Route Context (for rule evaluation)
// =============================================================================

/**
 * Context passed to routing rule evaluation.
 */
export interface RouteContext {
  /** Checkpoint name */
  checkpoint_name: string
  /** Run ID */
  run_id: string
  /** Checkpoint status */
  status: string
  /** Data asset */
  data_asset?: string
  /** Run time (ISO string) */
  run_time?: string
  /** Total issue count */
  total_issues?: number
  /** Critical issue count */
  critical_issues?: number
  /** High issue count */
  high_issues?: number
  /** Medium issue count */
  medium_issues?: number
  /** Low issue count */
  low_issues?: number
  /** Info issue count */
  info_issues?: number
  /** Pass rate (0-100) */
  pass_rate?: number
  /** Tags */
  tags?: Record<string, string>
  /** Metadata */
  metadata?: Record<string, unknown>
  /** Validation duration (ms) */
  validation_duration_ms?: number
  /** Error message */
  error?: string | null
}

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Get rule type info.
 */
export function getRuleTypeInfo(type: RoutingRuleType): RuleTypeInfo {
  return RULE_TYPE_INFO[type]
}

/**
 * Get rules by category.
 */
export function getRulesByCategory(category: 'simple' | 'compound' | 'expression'): RuleTypeInfo[] {
  return Object.values(RULE_TYPE_INFO).filter(info => info.category === category)
}

/**
 * Create a default rule config for a type.
 */
export function createDefaultRuleConfig(type: RoutingRuleType): RoutingRule {
  switch (type) {
    case 'always':
      return { type: 'always' }
    case 'never':
      return { type: 'never' }
    case 'status':
      return { type: 'status', statuses: ['failure'] }
    case 'severity':
      return { type: 'severity', min_severity: 'high' }
    case 'issue_count':
      return { type: 'issue_count', min_issues: 1 }
    case 'pass_rate':
      return { type: 'pass_rate', max_rate: 90 }
    case 'tag':
      return { type: 'tag', tags: {} }
    case 'data_asset':
      return { type: 'data_asset', pattern: '*' }
    case 'metadata':
      return { type: 'metadata', key_path: '' }
    case 'time_window':
      return { type: 'time_window', start_time: '09:00', end_time: '17:00' }
    case 'error':
      return { type: 'error' }
    case 'all_of':
      return { type: 'all_of', rules: [] }
    case 'any_of':
      return { type: 'any_of', rules: [] }
    case 'not':
      return { type: 'not', rule: { type: 'always' } }
    case 'jinja2':
      return { type: 'jinja2', expression: 'failed' }
    case 'python':
      return { type: 'python', expression: '' }
    default:
      return { type: 'always' }
  }
}

/**
 * Convert a rule to a human-readable description.
 */
export function describeRule(rule: RoutingRule): string {
  switch (rule.type) {
    case 'always':
      return 'Always'
    case 'never':
      return 'Never'
    case 'status':
      return `Status in [${rule.statuses.join(', ')}]${rule.negate ? ' (inverted)' : ''}`
    case 'severity':
      return `Severity >= ${rule.min_severity}${rule.min_count ? ` (min ${rule.min_count})` : ''}`
    case 'issue_count':
      return `Issues ${rule.min_issues ? `>= ${rule.min_issues}` : ''}${rule.max_issues ? ` <= ${rule.max_issues}` : ''}`
    case 'pass_rate':
      return `Pass rate ${rule.min_rate ? `>= ${rule.min_rate}%` : ''}${rule.max_rate ? ` <= ${rule.max_rate}%` : ''}`
    case 'tag':
      return `Tags: ${Object.entries(rule.tags).map(([k, v]) => v ? `${k}=${v}` : k).join(', ')}`
    case 'data_asset':
      return `Asset matches "${rule.pattern}"`
    case 'time_window':
      return `Time: ${rule.start_time || '00:00'} - ${rule.end_time || '23:59'}`
    case 'error':
      return rule.pattern ? `Error matches "${rule.pattern}"` : 'Has error'
    case 'all_of':
      return `All of (${rule.rules.length} rules)`
    case 'any_of':
      return `Any of (${rule.rules.length} rules)`
    case 'not':
      return `Not (${describeRule(rule.rule)})`
    case 'jinja2':
      return `Expression: ${rule.expression}`
    case 'metadata':
      return `Metadata: ${rule.key_path} ${rule.comparator || '=='} ${rule.expected_value}`
    case 'python':
      return `Python: ${rule.expression.substring(0, 50)}...`
    default:
      return 'Unknown rule'
  }
}

/**
 * Validate a route configuration.
 */
export function validateRoute(route: Route): string[] {
  const errors: string[] = []

  if (!route.name?.trim()) {
    errors.push('Route name is required')
  }

  if (!route.rule) {
    errors.push('Route rule is required')
  }

  if (!route.actions || route.actions.length === 0) {
    errors.push('At least one action is required')
  }

  return errors
}

File: src/truthound_dashboard/core/interfaces/routing.py
"""Routing interfaces for checkpoint-based validation pipelines.

Routing rules determine which actions to execute based on validation
results. They enable complex conditional logic for notifications and
post-validation processing.

This module defines abstract interfaces for routing that are loosely
coupled from truthound's checkpoint.routing module.

Routing features:
- Jinja2-based rule expressions
- Compound rules (AllOf, AnyOf, Not)
- Priority-based routing
- Action fanout (parallel execution)
- Context-based routing
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from enum import Enum
from typing import TYPE_CHECKING, Any, Protocol, runtime_checkable

if TYPE_CHECKING:
    from truthound_dashboard.core.interfaces.actions import (
        ActionContext,
        ActionProtocol,
        ActionResult,
    )
    from truthound_dashboard.core.interfaces.checkpoint import CheckpointResult


class RouteMode(str, Enum):
    """How routes are evaluated."""

    FIRST_MATCH = "first_match"  # Stop at first matching route
    ALL_MATCHES = "all_matches"  # Execute all matching routes
    PRIORITY = "priority"  # Execute in priority order


class RoutePriority(int, Enum):
    """Priority levels for routes."""

    CRITICAL = 100
    HIGH = 75
    MEDIUM = 50
    LOW = 25
    DEFAULT = 0


@dataclass
class RouteContext:
    """Context passed to routing rule evaluation.

    Attributes:
        checkpoint_result: The validation result.
        run_id: Unique run identifier.
        checkpoint_name: Name of the checkpoint.
        tags: Tags from the checkpoint.
        metadata: Additional metadata.
        variables: Custom variables for rule evaluation.
    """

    checkpoint_result: "CheckpointResult"
    run_id: str
    checkpoint_name: str
    tags: dict[str, str] = field(default_factory=dict)
    metadata: dict[str, Any] = field(default_factory=dict)
    variables: dict[str, Any] = field(default_factory=dict)

    def to_template_context(self) -> dict[str, Any]:
        """Convert to Jinja2 template context.

        Returns:
            Dictionary suitable for Jinja2 rendering.
        """
        result = self.checkpoint_result
        return {
            # Result properties
            "status": result.status.value,
            "passed": result.status.value == "success",
            "failed": result.status.value == "failure",
            "has_critical": getattr(result, "has_critical", False),
            "has_high": getattr(result, "has_high", False),
            "issue_count": getattr(result, "issue_count", 0),
            "critical_count": getattr(result, "critical_count", 0),
            "high_count": getattr(result, "high_count", 0),
            "medium_count": getattr(result, "medium_count", 0),
            "low_count": getattr(result, "low_count", 0),
            "row_count": getattr(result, "row_count", 0),
            "column_count": getattr(result, "column_count", 0),
            # Context properties
            "run_id": self.run_id,
            "checkpoint_name": self.checkpoint_name,
            "tags": self.tags,
            "metadata": self.metadata,
            # Custom variables
            **self.variables,
        }


@runtime_checkable
class RoutingRuleProtocol(Protocol):
    """Protocol for routing rule implementations.

    Routing rules evaluate checkpoint results and return True
    if the associated actions should be executed.

    Example:
        class SeverityRule:
            def __init__(self, min_severity: str):
                self.min_severity = min_severity

            def evaluate(self, context: RouteContext) -> bool:
                return context.checkpoint_result.has_severity(self.min_severity)
    """

    @property
    def name(self) -> str:
        """Get rule name."""
        ...

    @property
    def expression(self) -> str:
        """Get the rule expression (for Jinja2 rules)."""
        ...

    def evaluate(self, context: RouteContext) -> bool:
        """Evaluate the rule against the context.

        Args:
            context: Routing context with checkpoint result.

        Returns:
            True if rule matches (actions should execute).
        """
        ...


class BaseRoutingRule(ABC):
    """Abstract base class for routing rules.

    Provides common functionality for all routing rules.
    Subclasses must implement the evaluate method.
    """

    def __init__(
        self,
        name: str = "",
        description: str = "",
    ) -> None:
        """Initialize rule.

        Args:
            name: Rule name.
            description: Rule description.
        """
        self._name = name or self.__class__.__name__
        self._description = description

    @property
    def name(self) -> str:
        """Get rule name."""
        return self._name

    @property
    def expression(self) -> str:
        """Get the rule expression."""
        return ""

    @property
    def description(self) -> str:
        """Get rule description."""
        return self._description

    @abstractmethod
    def evaluate(self, context: RouteContext) -> bool:
        """Evaluate the rule."""
        ...


class Jinja2Rule(BaseRoutingRule):
    """Jinja2 expression-based routing rule.

    Evaluates a Jinja2 expression against the routing context.
    The expression should evaluate to a boolean.

    Example:
        rule = Jinja2Rule("critical_alert", "has_critical or critical_count > 0")
        if rule.evaluate(context):
            # Send critical alert
    """

    def __init__(
        self,
        name: str,
        expression: str,
        description: str = "",
    ) -> None:
        """Initialize Jinja2 rule.

        Args:
            name: Rule name.
            expression: Jinja2 expression that evaluates to boolean.
            description: Rule description.
        """
        super().__init__(name=name, description=description)
        self._expression = expression
        self._compiled: Any | None = None

    @property
    def expression(self) -> str:
        """Get the Jinja2 expression."""
        return self._expression

    def evaluate(self, context: RouteContext) -> bool:
        """Evaluate the Jinja2 expression.

        Args:
            context: Routing context.

        Returns:
            True if expression evaluates to truthy.
        """
        try:
            from jinja2 import Environment
        except ImportError:
            # Fallback to simple eval for basic expressions
            return self._fallback_evaluate(context)

        env = Environment()
        template_str = "{{ " + self._expression + " }}"
        template = env.from_string(template_str)
        result = template.render(**context.to_template_context())
        return result.lower() in ("true", "1", "yes")

    def _fallback_evaluate(self, context: RouteContext) -> bool:
        """Fallback evaluation without Jinja2."""
        ctx = context.to_template_context()

        # Handle simple expressions
        expr = self._expression.strip()

        # Direct variable lookup
        if expr in ctx:
            return bool(ctx[expr])

        # Simple comparisons
        for op in [" > ", " >= ", " < ", " <= ", " == ", " != "]:
            if op in expr:
                left, right = expr.split(op, 1)
                left_val = ctx.get(left.strip(), left.strip())
                right_val = ctx.get(right.strip(), right.strip())

                try:
                    left_val = float(left_val) if not isinstance(left_val, bool) else left_val
                    right_val = float(right_val) if not isinstance(right_val, bool) else right_val
                except (ValueError, TypeError):
                    pass

                if op == " > ":
                    return left_val > right_val
                elif op == " >= ":
                    return left_val >= right_val
                elif op == " < ":
                    return left_val < right_val
                elif op == " <= ":
                    return left_val <= right_val
                elif op == " == ":
                    return left_val == right_val
                elif op == " != ":
                    return left_val != right_val

        # Boolean operations
        if " or " in expr:
            parts = expr.split(" or ")
            return any(bool(ctx.get(p.strip(), False)) for p in parts)

        if " and " in expr:
            parts = expr.split(" and ")
            return all(bool(ctx.get(p.strip(), False)) for p in parts)

        return False


class AllOf(BaseRoutingRule):
    """Compound rule that matches when ALL child rules match.

    Example:
        rule = AllOf([
            Jinja2Rule("critical", "has_critical"),
            Jinja2Rule("high_count", "high_count > 10"),
        ])
    """

    def __init__(
        self,
        rules: list[BaseRoutingRule],
        name: str = "",
        description: str = "",
    ) -> None:
        """Initialize AllOf rule.

        Args:
            rules: Child rules that must all match.
            name: Rule name.
            description: Rule description.
        """
        super().__init__(name=name or "AllOf", description=description)
        self._rules = rules

    @property
    def rules(self) -> list[BaseRoutingRule]:
        """Get child rules."""
        return self._rules

    @property
    def expression(self) -> str:
        """Get combined expression."""
        exprs = [r.expression for r in self._rules if r.expression]
        return " and ".join(f"({e})" for e in exprs)

    def evaluate(self, context: RouteContext) -> bool:
        """Evaluate all child rules.

        Returns True only if all rules match.
        """
        return all(rule.evaluate(context) for rule in self._rules)


class AnyOf(BaseRoutingRule):
    """Compound rule that matches when ANY child rule matches.

    Example:
        rule = AnyOf([
            Jinja2Rule("critical", "has_critical"),
            Jinja2Rule("error", "status == 'error'"),
        ])
    """

    def __init__(
        self,
        rules: list[BaseRoutingRule],
        name: str = "",
        description: str = "",
    ) -> None:
        """Initialize AnyOf rule.

        Args:
            rules: Child rules (any one can match).
            name: Rule name.
            description: Rule description.
        """
        super().__init__(name=name or "AnyOf", description=description)
        self._rules = rules

    @property
    def rules(self) -> list[BaseRoutingRule]:
        """Get child rules."""
        return self._rules

    @property
    def expression(self) -> str:
        """Get combined expression."""
        exprs = [r.expression for r in self._rules if r.expression]
        return " or ".join(f"({e})" for e in exprs)

    def evaluate(self, context: RouteContext) -> bool:
        """Evaluate all child rules.

        Returns True if any rule matches.
        """
        return any(rule.evaluate(context) for rule in self._rules)


class NotRule(BaseRoutingRule):
    """Compound rule that inverts another rule.

    Example:
        rule = NotRule(Jinja2Rule("success", "passed"))
        # Matches when validation did NOT pass
    """

    def __init__(
        self,
        rule: BaseRoutingRule,
        name: str = "",
        description: str = "",
    ) -> None:
        """Initialize Not rule.

        Args:
            rule: Rule to invert.
            name: Rule name.
            description: Rule description.
        """
        super().__init__(name=name or "Not", description=description)
        self._rule = rule

    @property
    def rule(self) -> BaseRoutingRule:
        """Get the inverted rule."""
        return self._rule

    @property
    def expression(self) -> str:
        """Get negated expression."""
        return f"not ({self._rule.expression})"

    def evaluate(self, context: RouteContext) -> bool:
        """Evaluate the inverted rule.

        Returns True if the child rule does NOT match.
        """
        return not self._rule.evaluate(context)


class AlwaysRule(BaseRoutingRule):
    """Rule that always matches."""

    def __init__(self, name: str = "always") -> None:
        super().__init__(name=name, description="Always matches")

    @property
    def expression(self) -> str:
        return "True"

    def evaluate(self, context: RouteContext) -> bool:
        return True


class NeverRule(BaseRoutingRule):
    """Rule that never matches."""

    def __init__(self, name: str = "never") -> None:
        super().__init__(name=name, description="Never matches")

    @property
    def expression(self) -> str:
        return "False"

    def evaluate(self, context: RouteContext) -> bool:
        return False


# =============================================================================
# Route Definition
# =============================================================================


@dataclass
class Route:
    """A route defines a rule and its associated actions.

    When the rule matches, the actions are executed.

    Attributes:
        name: Route name for identification.
        rule: Routing rule to evaluate.
        actions: Actions to execute when rule matches.
        priority: Priority for route ordering.
        enabled: Whether this route is enabled.
        metadata: Additional metadata.
        stop_on_match: Stop evaluating other routes after this one matches.
    """

    name: str
    rule: BaseRoutingRule
    actions: list[str]  # Action names
    priority: RoutePriority = RoutePriority.DEFAULT
    enabled: bool = True
    metadata: dict[str, Any] = field(default_factory=dict)
    stop_on_match: bool = False

    def evaluate(self, context: RouteContext) -> bool:
        """Evaluate the route's rule.

        Args:
            context: Routing context.

        Returns:
            True if route matches.
        """
        if not self.enabled:
            return False
        return self.rule.evaluate(context)

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary."""
        return {
            "name": self.name,
            "rule_expression": self.rule.expression,
            "actions": self.actions,
            "priority": self.priority.value,
            "enabled": self.enabled,
            "metadata": self.metadata,
            "stop_on_match": self.stop_on_match,
        }


# =============================================================================
# Router Protocol
# =============================================================================


@runtime_checkable
class RouterProtocol(Protocol):
    """Protocol for router implementations.

    Routers evaluate routes and determine which actions to execute.

    Example:
        router = Router(mode=RouteMode.FIRST_MATCH)
        router.add_route(Route(
            name="critical_alert",
            rule=Jinja2Rule("critical", "has_critical"),
            actions=["pagerduty", "slack_critical"],
        ))

        matched_actions = router.route(context)
    """

    @property
    def mode(self) -> RouteMode:
        """Get routing mode."""
        ...

    @property
    def routes(self) -> list[Route]:
        """Get all routes."""
        ...

    def add_route(self, route: Route) -> None:
        """Add a route to the router.

        Args:
            route: Route to add.
        """
        ...

    def remove_route(self, name: str) -> bool:
        """Remove a route by name.

        Args:
            name: Route name.

        Returns:
            True if route was removed.
        """
        ...

    def route(self, context: RouteContext) -> list[str]:
        """Evaluate routes and return matching action names.

        Args:
            context: Routing context.

        Returns:
            List of action names to execute.
        """
        ...


class Router:
    """Default router implementation.

    Evaluates routes based on the configured mode and returns
    the actions that should be executed.
    """

    def __init__(
        self,
        mode: RouteMode = RouteMode.ALL_MATCHES,
        routes: list[Route] | None = None,
    ) -> None:
        """Initialize router.

        Args:
            mode: Routing mode.
            routes: Initial routes.
        """
        self._mode = mode
        self._routes: list[Route] = routes or []

    @property
    def mode(self) -> RouteMode:
        """Get routing mode."""
        return self._mode

    @property
    def routes(self) -> list[Route]:
        """Get all routes."""
        return self._routes.copy()

    def add_route(self, route: Route) -> None:
        """Add a route."""
        self._routes.append(route)
        # Re-sort by priority if in priority mode
        if self._mode == RouteMode.PRIORITY:
            self._routes.sort(key=lambda r: r.priority.value, reverse=True)

    def remove_route(self, name: str) -> bool:
        """Remove a route by name."""
        for i, route in enumerate(self._routes):
            if route.name == name:
                del self._routes[i]
                return True
        return False

    def route(self, context: RouteContext) -> list[str]:
        """Evaluate routes and return matching action names.

        Args:
            context: Routing context.

        Returns:
            List of unique action names to execute.
        """
        actions: list[str] = []
        seen: set[str] = set()

        # Sort routes if in priority mode
        routes = self._routes
        if self._mode == RouteMode.PRIORITY:
            routes = sorted(routes, key=lambda r: r.priority.value, reverse=True)

        for route in routes:
            if route.evaluate(context):
                for action in route.actions:
                    if action not in seen:
                        actions.append(action)
                        seen.add(action)

                # Stop if configured or first match mode
                if route.stop_on_match or self._mode == RouteMode.FIRST_MATCH:
                    break

        return actions

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary."""
        return {
            "mode": self._mode.value,
            "routes": [r.to_dict() for r in self._routes],
        }
